### DISCUSSION TITLE
Cannot get XML node Value

### DISCUSSION THREAD

**Discussion by fabiodap** [nextgenhealthcare/connect#4817] [General]
Hello,

we are trying to read the content of a node.

We use the following piece of code:

     var msgNoNamespace= r.message.replace(' xmlns="urn:hl7-org:v2xml"', '');
     var parsed = new XML(msgNoNamespace);
     messageContent =parsed.*::['Body'].*::['sendMessageResponse'].*::['responseContent'];

     var msaStatus = messageContent['MSA']['MSA.6']['CE.1'];
     logger.info('msaStatus : '+msaStatus );

The result is that msaStatus is empty. I expected to get the value: "KO".

We start from the following Json. 
r.message is the content of "message".
We extract the HL7 message from the tag **responseContent**,
and then we try to read the value of: MSA/MSA.6/CE.1

{
  "results" : {
    "1111" : {
      "status" : "SENT",
      "message" : "<SOAP-ENV:Envelope xmlns:SOAP-ENV=\"http://schemas.xmlsoap.org/soap/envelope/\"><SOAP-ENV:Header/><SOAP-ENV:Body xmlns:ns2=\"urn:it.cup2000.cup.servizi\"><ns2:sendMessageResponse><ns2:status>OK</ns2:status><ns2:responseContent>&lt;?xml version=\"1.0\"?&gt;\r\n&lt;SRR_S01 xmlns=\"urn:hl7-org:v2xml\"&gt;\r\n    &lt;MSH&gt;\r\n        &lt;MSH.1&gt;|&lt;/MSH.1&gt;\r\n        &lt;MSH.2&gt;^~\\&amp;amp;&lt;/MSH.2&gt;\r\n        &lt;MSH.3&gt;\r\n            &lt;HD.1&gt;CUP2000&lt;/HD.1&gt;\r\n        &lt;/MSH.3&gt;\r\n        &lt;MSH.4&gt;\r\n            &lt;HD.1&gt;CUP2000&lt;/HD.1&gt;\r\n        &lt;/MSH.4&gt;\r\n        &lt;MSH.7&gt;\r\n            &lt;TS.1&gt;202110271727&lt;/TS.1&gt;\r\n        &lt;/MSH.7&gt;\r\n        &lt;MSH.9&gt;\r\n            &lt;MSG.1&gt;SRR&lt;/MSG.1&gt;\r\n            &lt;MSG.2&gt;S01&lt;/MSG.2&gt;\r\n            &lt;MSG.3&gt;SRR_S01&lt;/MSG.3&gt;\r\n        &lt;/MSH.9&gt;\r\n        &lt;MSH.10&gt;FB7A943D3C0BDF87AABB&lt;/MSH.10&gt;\r\n        &lt;MSH.11&gt;\r\n            &lt;PT.1&gt;P&lt;/PT.1&gt;\r\n        &lt;/MSH.11&gt;\r\n        &lt;MSH.12&gt;\r\n            &lt;VID.1&gt;2.5&lt;/VID.1&gt;\r\n        &lt;/MSH.12&gt;\r\n        &lt;MSH.13&gt;1&lt;/MSH.13&gt;\r\n    &lt;/MSH&gt;\r\n    &lt;MSA&gt;\r\n        &lt;MSA.1&gt;AE&lt;/MSA.1&gt;\r\n        &lt;MSA.6&gt;\r\n            &lt;CE.1&gt;KO&lt;/CE.1&gt;\r\n        &lt;/MSA.6&gt;\r\n    &lt;/MSA&gt;\r\n    &lt;ERR&gt;\r\n        &lt;ERR.3&gt;\r\n            &lt;CWE.1&gt;207&lt;/CWE.1&gt;\r\n            &lt;CWE.2&gt;Application internal error&lt;/CWE.2&gt;\r\n        &lt;/ERR.3&gt;\r\n        &lt;ERR.4&gt;E&lt;/ERR.4&gt;\r\n        &lt;ERR.5&gt;\r\n            &lt;CWE.1&gt;APP.102&lt;/CWE.1&gt;\r\n            &lt;CWE.2&gt;Errore durante il parsing del messaggio ricevuto&lt;/CWE.2&gt;\r\n        &lt;/ERR.5&gt;\r\n        &lt;ERR.8&gt;APP.102 - Errore durante il parsing del messaggio ricevuto&lt;/ERR.8&gt;\r\n    &lt;/ERR&gt;\r\n&lt;/SRR_S01&gt;\r\n</ns2:responseContent><ns2:responseVersion>1.0.000</ns2:responseVersion><ns2:responseDate>2021-10-27T17:27:00.984+02:00</ns2:responseDate></ns2:sendMessageResponse></SOAP-ENV:Body></SOAP-ENV:Envelope>",
      "duration" : "PT0.355S",
      "startTime" : "2021-10-27T17:27:00.511"
    }
  },
  "statuses" : {
    "SENT" : 2
  },
  "runDuration" : "PT0.386S"
}

This is the content of "messageContent", stripped of the name space declaration.:

<?xml version="1.0"?>
<SRR_S01>
    <MSH>
        <MSH.1>|</MSH.1>
        <MSH.2>^~\&amp;</MSH.2>
        <MSH.3>
            <HD.1>CUP2000</HD.1>
        </MSH.3>
        <MSH.4>
            <HD.1>CUP2000</HD.1>
        </MSH.4>
        <MSH.7>
            <TS.1>202110271727</TS.1>
        </MSH.7>
        <MSH.9>
            <MSG.1>SRR</MSG.1>
            <MSG.2>S01</MSG.2>
            <MSG.3>SRR_S01</MSG.3>
        </MSH.9>
        <MSH.10>65971446D96FC92E852D</MSH.10>
        <MSH.11>
            <PT.1>P</PT.1>
        </MSH.11>
        <MSH.12>
            <VID.1>2.5</VID.1>
        </MSH.12>
        <MSH.13>1</MSH.13>
    </MSH>
    <MSA>
        <MSA.1>AE</MSA.1>
        <MSA.6>
            <CE.1>KO</CE.1>
        </MSA.6>
    </MSA>
    <ERR>
        <ERR.3>
            <CWE.1>207</CWE.1>
            <CWE.2>Application internal error</CWE.2>
        </ERR.3>
        <ERR.4>E</ERR.4>
        <ERR.5>
            <CWE.1>APP.102</CWE.1>
            <CWE.2>Errore durante il parsing del messaggio ricevuto</CWE.2>
        </ERR.5>
        <ERR.8>APP.102 - Errore durante il parsing del messaggio ricevuto</ERR.8>
    </ERR>
</SRR_S01>

greetings,
Fabio

**Comment by pacmano1** [comment]
How about posting valid JSON first?

**Reply by pacmano1** [reply]
Can you share the entire channel?   It's puzzling to see  HL7 in SOAP in JSON.

**Comment by fabiodap** [comment]
Hello,

problem solved.

It was enough to add a .toString here:

     messageContent =parsed.*::['Body'].*::['sendMessageResponse'].*::['responseContent'].toString();
     logger.info('messageContent: '+messageContent);
     
     var messageContentParsed = new XML(messageContent);

greetings,
Fabio
